articles/working-with-pythons-unittest/unittests.md

```diff
@@ -0,0 +1,216 @@
+```
+
+Of the 3 tests that we performed, `self.assertEqual(double(8), 16)` produces an error since `32 != 16`. This demonstrates how unit tests are used for regression testing.
+
+To see more detailed information, use the `-v` flag before the file holding the unit tests. The `v` stands for `verbose`.
+
+```console
+>>> python3 -m unittest -v main.py
+test_divide (tests.TestMathFunctions) ... ok
+test_double (tests.TestMathFunctions) ... FAIL
+test_square (tests.TestMathFunctions) ... ok
+
+======================================================================
+FAIL: test_double (tests.TestMathFunctions)
+----------------------------------------------------------------------
+Traceback (most recent call last):
+  File "main.py", line 8, in test_double
+    self.assertEqual(double(8), 16)
+AssertionError: 32 != 16
+
+----------------------------------------------------------------------
+Ran 3 tests in 0.001s
+
+FAILED (failures=1)
+```
+
+##### Error messages
+Error messages are very useful for future code debugging. They also provide documentation for quality assurance testers and other developers.
+For instance, to add an error message to `self.assertEqual(double(8), 16)`, change your code to match this:
+
+```python
+   def test_double(self):
+      self.assertEqual(double(8), 16, "the function should return two times the number provided")
+```
+
+The output will be:
+
+```console
+>>> python3 -m unittest main.py
+.F.
+======================================================================
+FAIL: test_double (__main__.TestMathFunctions)
+----------------------------------------------------------------------
+Traceback (most recent call last):
+  File "main.py", line 8, in test_double
+    self.assertEqual(double(8), 16, "the function should return two times the number provided")
+AssertionError: 32 != 16 : the function should return two times the number provided
+
+----------------------------------------------------------------------
+Ran 3 tests in 0.001s
+
+FAILED (failures=1)
+```
+
+As demonstrated above, the error message we added describes the `AssertionError` better. This has made it more understandable.
+
+##### More assert methods
+So far, we've seen the `assertEqual(a, b)` method, the `assertNotEqual(a, b)` method, and the `assertRaises(x)` method. Apart from these, the unittest module provides extra assert methods which include:
+
+1. `assertTrue(x)` checks that a boolean value `x` equals `True`.
+2. `assertFalse(x)` checks that a boolean value `x` equals `False`.
+3. `assertIn(a,b)` checks that a value `a` exists in a list `b`. For instance, this check should pass when `a=2` and `b=[1,2,3]`.
+4. `asserNottIn(a,b)` checks that a value `a` does not exist in a list `b`. For instance, this check should fail when `a=2` and `b=[1,2,3]` since `2` exists in the list `b`.
+
+Once you are comfortable working with these, you can check more assert methods [here](https://docs.python.org/3/library/unittest.html#assert-methods).
+
```

Comment:
Can you add a table contains all the assert methods here instead of this?
You can use [this tool](https://www.tablesgenerator.com/markdown_tables).
For example:
|  Method |  Result |   
|---|---|
|  assertTrue(x) |  checks that a boolean value `x` equals `True` | 
|  assertFalse(x) |  checks that a boolean value `x` equals `False`  |  

In Section, we ensure the reader can get everything about the topic inside the article itself.